rhizome/kubernetes/lib/ubi_cni.rb

```diff
@@ -49,189 +28,199 @@ def run
     r "echo 1 > /proc/sys/net/ipv4/conf/#{outer_ifname}/proxy_arp"
-
-    {container_ip: container_ip, gateway_ip: gateway_ip}
   end
 
   def handle_del
+    check_required_env_vars(["CNI_CONTAINERID"])
     container_id = ENV["CNI_CONTAINERID"]
-
-    if @ipam_store["allocated_ips"].key?(container_id)
-      @ipam_store["allocated_ips"].delete(container_id)
-      save_ipam_store
+    @logger.info "Releasing IPs for container #{container_id}"
+    with_ipam_store do |ipam_store|
+      ipam_store["allocated_ips"].delete(container_id)
     end
-
     "{}"
   end
 
   def handle_get
+    check_required_env_vars(["CNI_NETNS", "CNI_IFNAME"])
     cni_netns = ENV["CNI_NETNS"].sub("/var/run/netns/", "")
     inner_ifname = ENV["CNI_IFNAME"]
 
+    @logger.info "Retrieving configuration for interface #{inner_ifname} in namespace #{cni_netns}"
     inner_mac = r("ip -n #{cni_netns} link show #{inner_ifname}").match(/link\/ether ([0-9a-f:]+)/)[1]
     container_ip = r("ip -n #{cni_netns} -6 addr show dev #{inner_ifname}").match(/inet6 ([0-9a-f:\/]+)/)[1]
 
     dns_config_path = "/etc/netns/#{cni_netns}/resolv.conf"
     dns_servers = []
     search_domains = []
     if File.exist?(dns_config_path)
-      File.readlines(dns_config_path).each do |line|
-        if line.start_with?("nameserver")
-          dns_servers << line.split[1]
-        elsif line.start_with?("search")
-          search_domains = line.split.drop(1)
-        end
+      File.readlines(dns_config_path, chomp: true).each do |line|
+        dns_servers << line.split[1] if line.start_with?("nameserver")
+        search_domains = line.split.drop(1) if line.start_with?("search")
       end
     end
 
     response = {
       cniVersion: "1.0.0",
-      interfaces: [
-        {
-          name: inner_ifname,
-          mac: inner_mac,
-          sandbox: "/var/run/netns/#{cni_netns}"
-        }
-      ],
-      ips: [
-        {
-          address: container_ip,
-          gateway: nil,
-          interface: 0
-        }
-      ],
-      dns: {
-        nameservers: dns_servers,
-        search: search_domains,
-        options: ["ndots:5"]
-      }
+      interfaces: [{name: inner_ifname, mac: inner_mac, sandbox: "/var/run/netns/#{cni_netns}"}],
+      ips: [{address: container_ip, gateway: nil, interface: 0}],
+      dns: {nameservers: dns_servers, search: search_domains, options: ["ndots:5"]}
     }
-
     JSON.generate(response)
   end
 
+  def with_ipam_store
+    lock_file_path = "#{IPAM_STORE_FILE}.lock"
+    File.open(lock_file_path, File::RDWR | File::CREAT, 0o644) do |lock_file|
+      lock_file.flock(File::LOCK_EX)
+      begin
+        content = File.read(IPAM_STORE_FILE)
+        ipam_store = content.empty? ? {"allocated_ips" => {}} : JSON.parse(content)
+      rescue Errno::ENOENT
+        ipam_store = {"allocated_ips" => {}}
+      end
+      result = yield ipam_store
+      safe_write_to_file(IPAM_STORE_FILE, JSON.pretty_generate(ipam_store))
```

Comment:
I'm a bit confused about why with_ipam_store has to flock but also safe_write_to_file, perhaps safe_write_to_file needs to get a little smarter or refactored so other stuff can be done while it holds its lock. the code seems nearly in duplicate, but I suppose some kind of compositional weakness of safe_write_to_file is why you're doing that?